docs: release notes for 1.7.0 (#776)
canonical/rockcraft: docs/release-notes/rockcraft-1-7-0.rst

Author: https://canonical-ubuntu-documentation-library--776.com.readthedocs.build/rockcraft/en/776/release-notes/rockcraft-1-7-0/

.. _release-1.7.0:

> Rockcraft 1.7.0 release notes

Author: @medubelko as we discussed earlier, I ended up putting everything that I would put on a changelog here. Please take a look and point out the items you feel are out-of-place here.

> Rockcraft 1.7.0 release notes

Author: I also removed some sections, like the "Requirements" one, which would be empty for this particular release.

Reviewer: It was decided that the _Requirements_ section should be mandatory for each page, so it will have to stay in. It's in effect boilerplate.

Author: Alright, but in this case we should review the content - the default template lists "Python 3.x or higher", but in the case of the applications the "official" means of distribution is the snap. In this case, there's no requirement on the user even having Python, but they do need a snap-capable system. Shall I draft a list of requirements here that we can possibly use for the other applications too?

Reviewer: Yes, please! Feel free to crib from Snapcraft's system reqs, too: https://canonical-snapcraft.readthedocs-hosted.com/en/latest/reference/system-requirements/

Author: done! Thanks for the cheatsheet

=============================

20 December 2024

Learn about the new features, changes, and fixes introduced in Rockcraft 1.7.0.
For information about the Rockcraft release cycle, see the
:ref:`release_policy_and_schedule`.

What's new
----------

Rockcraft 1.7.0 brings the following features, integrations, and improvements.

Support for remote builds
~~~~~~~~~~~~~~~~~~~~~~~~~

We've added support for remote builds, which are powered by Canonical's
Launchpad platform and allow you to build rocks for multiple architectures with
a single command.

This feature is experimental and subject to change. Feel free to provide

> feedback on the `Ubuntu Discourse`_ forum.

Author: Ah I see. I have a task to document the feature on craft-application, but it won't be ready & integrated into Rockcraft by the end of this week. So I added a call to run ``rockcraft remote-build --help``, which has more information and hopefully the user can take it from there.

> feedback on the `Ubuntu Discourse`_ forum.

Reviewer: Perfect!

Minor features
--------------

Rockcraft 1.7.0 brings the following minor changes.

> "init" command

Author: I don't know where I got this from but I thought we preferred quotes over the inline-code markers. Guess I was wrong!

> "init" command

Reviewer: Not from me. :hear_no_evil:

~~~~~~~~~~~~~~

- The default base of the project file generated by the "rockcraft init" command has

>   been updated from "ubuntu@22.04" to "ubuntu@24.04".

Author: maybe this was my confusion: should these be `"ubuntu@22.04"` or inline-code-formatted?

Reviewer: One more thought about inline code, while I'm thinking about this:

Sometimes a symbol can become a real word. For example, core22 and commands like grep, which were originally intended as machine expressions, have entered general parlance. Things like that, which have become actual words everyday speech, don't need code fences. Determining whether something is a word can sometimes be difficult, so you can leave it up to me.

Rockcraft snap
~~~~~~~~~~~~~~

- The Rockcraft snap has been updated from core22 to core24.

Go extension
~~~~~~~~~~~~

> - The documentation for the "go-framework" extension has been updated with
>   improved tutorial and reference.

Reviewer: ```suggestion
A new :ref:`tutorial <build-a-rock-for-a-go-application>` and :ref:`reference
<go-framework-reference>` are available for the go-framework extension.
```

OCI image configuration
~~~~~~~~~~~~~~~~~~~~~~~

- When using "run-user: _daemon_", the value of the "Config.User" key in the
  generated rock has been updated from the name ("_daemon_") to the numeric id.

12-factor documentation
~~~~~~~~~~~~~~~~~~~~~~~

- The documentation for the various 12-factor framework extensions has been
  updated for clarity and to better conform with Canonical's guidelines.

Maven and Ant plugins
~~~~~~~~~~~~~~~~~~~~~

> - Updated the Maven and Ant plugins to generate the "/usr/bin/java" path to the

Author: I updated this paragraph to try to address both concerns (this and the agentless passive voice) - please take a look

> - Updated the Maven and Ant plugins to generate the "/usr/bin/java" path to the

Reviewer: The new copy makes much more sense, thanks!

  Java runtime executable, instead of the "/bin/java" symlink which conflicts

>   with the user-merged Ubuntu base.

Author: I had something similar to this originally, but the template points to using the past tense. I can see how it could depend on the specific instance, but won't the notes look awkward if some are in the past tense and some use the present one?

Fixed bugs and issues
---------------------

> The following issues have been resolved in Rockcraft 1.7.1:

Author: ```suggestion
The following issues have been resolved in Rockcraft 1.7.0:
```

> - #659 Kubernetes ``runAsNonRoot`` incompatibility with ``run-user: _daemon_``.

Author: Sure - I think we need to make it clear-er in the template that these need to be links

- #749 Trying to build the output of "expand-extensions" fails because of the
  pebble part.

Contributors
------------

We would like to express a big thank you to all the people who contributed to
this release.

> `@dariuszd21 <https://github.com/dariuszd21>`_,

Author: the spellchecker is complaining about all of these, heh

Reviewer: Annoying. In the short-term, what would be easier?

- Use :literalref: instead
- Add a custom block directive that's a complement for the inline `:spellexception:`

Author: updated to use `:literalref:`

> `@dariuszd21 <https://github.com/dariuszd21>`_,

Reviewer: In terms of visual presentation, I'm not sure I like splitting the names across lines. I'd prefer one big paragraph.

Author: It's rendered as a single paragraph; they are split over sequential lines here because of the line-length limit, which basically enforces "one per line" in this case
![image](https://github.com/user-attachments/assets/e17eebc4-9ed4-4e52-af33-7fc4da407e6b)

Reviewer: Ah! I reacted instinctively to the code and not the preview, oops.

This feature is experimental and subject to change, and more information is
available by running ``rockcraft remote-build --help``. Feel free to provide
feedback on the `Ubuntu Discourse`_ forum.

Minor features
--------------

Rockcraft 1.7.0 brings the following minor changes.

``init`` command
~~~~~~~~~~~~~~~~

The default base of the project file generated by ``rockcraft init`` has been
updated from "ubuntu@22.04" to "ubuntu@24.04".

Rockcraft snap
~~~~~~~~~~~~~~

The base of the Rockcraft snap was updated from core22 to core24.

Go extension
~~~~~~~~~~~~

A new :ref:`tutorial <build-a-rock-for-a-go-application>` and :ref:`reference
<go-framework-reference>` are available for the go-framework extension.

OCI image configuration
~~~~~~~~~~~~~~~~~~~~~~~

When using ``run-user: _daemon_``, the value of the ``Config.User`` key in the
generated rock has been updated from the user name (``_daemon_``) to its
numeric ID.

12-factor documentation
~~~~~~~~~~~~~~~~~~~~~~~

The documentation for the various 12-factor framework extensions has been
updated for clarity and to better conform with documentation best practices.

Maven and Ant plugins
~~~~~~~~~~~~~~~~~~~~~

> Changed the location of the Java runtime executable on rocks built using the
> Maven and Ant plugins from ``/bin/java`` to ``/usr/bin/java``, as the former
> conflicts with the user-merged Ubuntu base.

Reviewer: I'm still not a fan of starting a sentence with a past tense verb, so let's insert ourselves:

```suggestion
We changed the location of the Java runtime executable on rocks built using the
the Maven and Ant plugins from ``/bin/java`` to ``/usr/bin/java``, as the
former conflicts with the user-merged Ubuntu base.
```

Author: I accepted this suggestion "manually", as I worked on the page here

=============================

20 December 2024

Learn about the new features, changes, and fixes introduced in Rockcraft 1.7.0.
For information about the Rockcraft release cycle, see the
:ref:`release_policy_and_schedule`.

Requirements and compatibility
------------------------------

To run Rockcraft, a system requires the following minimum hardware and
installed software. These requirements apply to local hosts as well as VMs and
container hosts.

Minimum hardware requirements
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- AMD64, ARM64, ARMv7-M, RISC-V 64-bit, PowerPC 64-bit little-endian, or S390x
  processor
- 2GB RAM
- 10GB available storage space
- Internet access for remote software sources and the Snap Store

Platform requirements
~~~~~~~~~~~~~~~~~~~~~

.. list-table::
  :header-rows: 1
  :widths: 1 3 3

  * - Platform
    - Version
    - Software requirements
  * - GNU/Linux
    - Popular distributions that ship with systemd and are `compatible with
      snapd <https://snapcraft.io/docs/installing-snapd>`_
    - systemd

What's new
----------

Rockcraft 1.7.0 brings the following features, integrations, and improvements.

Support for remote builds
~~~~~~~~~~~~~~~~~~~~~~~~~

> We added support for remote builds, which leverage off-site compute for

Author: @medubelko here's what I finally ended up with after much debating with myself

To run Rockcraft, a system requires the following minimum hardware and
installed software. These requirements apply to local hosts as well as VMs and
container hosts.

Minimum hardware requirements
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- AMD64, ARM64, ARMv7-M, RISC-V 64-bit, PowerPC 64-bit little-endian, or S390x
  processor
- 2GB RAM
- 10GB available storage space
- Internet access for remote software sources and the Snap Store

Platform requirements
~~~~~~~~~~~~~~~~~~~~~

.. list-table::
  :header-rows: 1
  :widths: 1 3 3

  * - Platform
    - Version
    - Software requirements
  * - GNU/Linux
    - Popular distributions that ship with systemd and are `compatible with
      snapd <https://snapcraft.io/docs/installing-snapd>`_
    - systemd

What's new
----------

Rockcraft 1.7.0 brings the following features, integrations, and improvements.

Support for remote builds
~~~~~~~~~~~~~~~~~~~~~~~~~

We added support for remote builds, which leverage off-site compute for
building rocks and can build for multiple architectures with a single command.
Remote builds are hosted on Launchpad.

This feature is experimental and subject to change, and more information is

> available by running ``rockcraft remote-build --help``. Feel free to provide

Author: Adding a note to myself here that once #778 is merged I can update these to point to the new docs

feedback on the `Ubuntu Discourse`_ forum.

Minor features
--------------

Rockcraft 1.7.0 brings the following minor changes.

``init`` command
~~~~~~~~~~~~~~~~

The default base of the project file generated by ``rockcraft init`` has been
updated from ``ubuntu@22.04`` to ``ubuntu@24.04``.

Rockcraft snap
~~~~~~~~~~~~~~

The base of the Rockcraft snap was updated from core22 to core24.

Go extension
~~~~~~~~~~~~

A new :ref:`tutorial <build-a-rock-for-a-go-application>` and :ref:`reference
<go-framework-reference>` are available for the go-framework extension.

OCI image configuration
~~~~~~~~~~~~~~~~~~~~~~~

When using ``run-user: _daemon_``, the value of the ``Config.User`` key in the
generated rock has been updated from the user name (``_daemon_``) to its
numeric ID.

> 12-factor extensions

Author: This entry is new, to reflect the PR that got merged yesterday

~~~~~~~~~~~~~~~~~~~~

> We added support for Gunicorn async workers on the Django and Flask 12-factor

Reviewer: > 12-factor

Should we call them "12-factor extensions"? "Django and Flask extensions, for supporting 12-factor apps" might be clearer.

Author: reworded this: kept as Django and Flask, and updated the title to `12-factor app extensions`

extensions.

> 12-factor documentation
> ~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
12-factor app documentation
~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
